rpcs3/Emu/Cell/PPUInterpreter.cpp

```diff
@@ -4673,36 +4648,59 @@ bool ppu_interpreter::FNMADDS(ppu_thread& ppu, ppu_opcode_t op)
 
 bool ppu_interpreter::MTFSB1(ppu_thread& ppu, ppu_opcode_t op)
 {
-	LOG_WARNING(PPU, "MTFSB1");
+	if (const u32 bit = op.crbd; true)
+	{
+		if (bit < 16 || bit > 19) LOG_WARNING(PPU, "MTFSB1(%d)", bit);
+		ppu.fpscr.bits[op.crbd] = 1;
+	}
+
 	if (UNLIKELY(op.rc)) ppu_cr_set(ppu, 1, ppu.fpscr.fg, ppu.fpscr.fl, ppu.fpscr.fe, ppu.fpscr.fu);
 	return true;
 }
 
 bool ppu_interpreter::MCRFS(ppu_thread& ppu, ppu_opcode_t op)
 {
-	LOG_WARNING(PPU, "MCRFS");
-	ppu_cr_set(ppu, op.crfd, false, false, false, false);
+	if (op.crfs != 4) LOG_WARNING(PPU, "MCRFS(%d)", op.crfs);
+	reinterpret_cast<u32*>(&ppu.cr)[op.crfd] = reinterpret_cast<u32*>(&ppu.fpscr.bits[0])[op.crfs];
 	return true;
 }
 
 bool ppu_interpreter::MTFSB0(ppu_thread& ppu, ppu_opcode_t op)
 {
-	LOG_WARNING(PPU, "MTFSB0");
+	if (const u32 bit = op.crbd; true)
+	{
+		if (bit < 16 || bit > 19) LOG_WARNING(PPU, "MTFSB0(%d)", bit);
+		ppu.fpscr.bits[op.crbd] = 0;
+	}
+
 	if (UNLIKELY(op.rc)) ppu_cr_set(ppu, 1, ppu.fpscr.fg, ppu.fpscr.fl, ppu.fpscr.fe, ppu.fpscr.fu);
 	return true;
 }
 
 bool ppu_interpreter::MTFSFI(ppu_thread& ppu, ppu_opcode_t op)
 {
-	LOG_WARNING(PPU, "MTFSFI");
+	if (u32 bf = op.crfd, i = op.i; true)
+	{
+		if (bf != 4)
+		{
+			LOG_WARNING(PPU, "MTFSFI(%d)", bf);
+			return true;
+		}
+
+		ppu.fpscr.fu = i & 1; i >>= 1;
+		ppu.fpscr.fe = i & 1; i >>= 1;
+		ppu.fpscr.fg = i & 1; i >>= 1;
+		ppu.fpscr.fl = i & 1;
+	}
+
 	if (UNLIKELY(op.rc)) ppu_cr_set(ppu, 1, ppu.fpscr.fg, ppu.fpscr.fl, ppu.fpscr.fe, ppu.fpscr.fu);
 	return true;
 }
 
 bool ppu_interpreter::MFFS(ppu_thread& ppu, ppu_opcode_t op)
 {
 	LOG_WARNING(PPU, "MFFS");
-	ppu.fpr[op.frd] = 0.0;
+	ppu.fpr[op.frd] = u64{ppu.fpscr.fl} << 15 | u64{ppu.fpscr.fg} << 14 | u64{ppu.fpscr.fe} << 13 | u64{ppu.fpscr.fu} << 12;
```

Comment:
Won't produce a correct result due to conversion from u64 to double.

Reply:
right.